File: i3-exec-wait.rb
#!/usr/bin/env ruby
#
# Execute a command and wait for i3 to register window(s).
# I know it is ugly.
#
# http://www.drbig.one.pl
#

%w{bindata json socket}.each {|g| require g }

# data structures
class I3Head < BinData::Record
  endian  :little
  int32   :len
  int32   :type
end

class I3Data < BinData::Record
  endian  :little
  string  :magic, :initial_value => 'i3-ipc'
  int32   :len, :value => lambda { payload.length }
  int32   :type
  string  :payload, :read_length => :len
end

# print help message
def help
  STDERR.puts 'Usage: i3-exec-wait.rb (-n <how_many_windows> | -r <regexp>) command'
  STDERR.puts 'Without options it will just wait for any window to register.'
  STDERR.puts 'The -n option will wait for a number of windows to register.'
  STDERR.puts 'With a regexp it will wait for a window whose _initial_ title matches the regexp - beware that initial title may be different from what you see (e.g. opening urxvt or xterm with shell that changes window title).'
  exit(2)
end

# parse an array to a command string preserving spaces
def parse_cmd(array)
  array.collect do |e|
    if e.chars.to_a.include? ' '
      '"' + e + '"'
    else
      e
    end
  end.join(' ')
end

# get a message from i3-ipc
# or bail put saying why
def get_msg(socket)
  unless socket.read(6) == 'i3-ipc'
    STDERR.puts 'Wrong magic in response, bye.'
    exit(2)
  end

  head = I3Head.read(socket.read(8))
  begin
    payload = JSON.parse(socket.read(head.len))
  rescue
    STDERR.puts 'Error on getting or parsing payload, bye.'
    exit(2)
  end

  payload
end

# handle arguments, what a mess
# try to be nice and say what's wrong
if ARGV.length == 0
  help
else
  mode = false
  nwin = false
  regexp = false
  cmd = false

  if %w{--help -help -h}.member? ARGV.first
    help
  elsif ARGV.first == '-n'
    mode = :nwin
    begin
      nwin = ARGV[1].to_i
    rescue
      STDERR.puts 'Argument to -n should be an integer, sorry.'
      exit(2)
    end
    if ARGV.length < 3
      STDERR.puts 'You need to specify a command...'
      exit(2)
    end
    cmd = parse_cmd(ARGV.slice(2, ARGV.length-1))
  elsif ARGV.first == '-r'
    mode = :regexp
    begin
      regexp = Regexp.new(ARGV[1])
    rescue
      STDERR.puts 'Argument to -r should be a valid (Ruby) regexp, sorry.'
    end
    if ARGV.length < 3
      STDERR.puts 'You need to specify a command...'
      exit(2)
    end
    cmd = parse_cmd(ARGV.slice(2, ARGV.length-1))
  else
    mode = :simple
    cmd = parse_cmd(ARGV)
  end
end

UNIXSocket.open(%x{i3 --get-socketpath}.chop) do |s|
  s.write(I3Data.new(:type => 2, :payload => %w{window}.to_json).to_binary_s)
  unless get_msg(s)['success']
    STDERR.puts 'Could not subscribe to window events, bye.'
    exit(2)
  end

  begin
    Process.detach(Process.spawn(cmd))
  rescue
    STDERR.puts 'Error executing your command, bye.'
    exit(2)
  end

  case mode
  when :simple
    get_msg(s)
  when :nwin
    while nwin > 0
      get_msg(s)
      nwin -= 1
    end
  when :regexp
    while not get_msg(s)['container']['name'].match regexp
      # nothing
    end
  end

  exit(0)
end

File: i3-exec-wait.c
/*
 * i3-exec-wait
 * © 2013 Piotr S. Staszewski
 * For licensing purposes assume current i3 license.
 *
 * http://www.drbig.one.pl
 *
 */

#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

#define SUBSCRIBE "i3-ipc\n\x00\x00\x00\x02\x00\x00\x00[\"window\"]"

/*
 * Print help.
 *
 */
void help() {
  fprintf(stderr, "Supply a command you wish to execute (and its arguments if needed).\n");
  fprintf(stderr, "If you need to wait for more than one window use -n <INT> as first two arguments.\n");
  fprintf(stderr, "E.g. i3-exec-wait xterm -title \"Rc Shell\" -e rc\n");
  fprintf(stderr, "     i3-exec-wait -n 4 gimp\n");
  exit(1);
}

/*
 * Print an error and exit.
 *
 */
void die(const char *msg) {
  if (errno) {
    perror(msg);
    exit(2);
  } else {
    fprintf(stderr, "ERROR: %s!\n", msg);
    exit(1);
  }
}

/*
 * Receive message from i3-ipc.
 * Returns pointer to payload or NULL.
 *
 */
char *get_msg(int sock) {
  // The i3-ipc message header, packing is very important.
  #pragma pack(1)
  struct ipc_header {
    uint8_t magic[6];
    uint32_t len;
    uint32_t type;
  } head;
  char *payload;
  int len;

  if ((len = recv(sock, &head, sizeof(head), MSG_WAITALL)) < 0)
    die("recv");

  if (head.len > 0) {
    if ((payload = calloc(head.len + 1, sizeof(uint8_t))) == NULL)
      die("calloc");
    if (recv(sock, payload, head.len, MSG_WAITALL) < 0)
      die("recv");
    return payload;
  } else
    return NULL;
}

int main(int argc, char *argv[]) {
  int nwin = 1;                 // for how many windows we wait.
  int ncmd = 1;                 // where do command argv starts.
  struct sockaddr_un ipcaddr;   // i3-ipc addr.
  int sock;                     // i3-ipc socket.

  /* Parse arguments and setup nwin and ncmd. */
  if (argc < 2)
    help();
  else {
    if (!strcmp(argv[1], "-h") || !strcmp(argv[1], "--help"))
      help();
    else if (!strcmp(argv[1], "-n")) {
      if (argc < 4)
        die("You have to specify number of windows and a command");
      nwin = atoi(argv[2]);
      if (nwin < 1)
        die("Number of windows has to be > 0");
      ncmd = 3;
    }
  }

  /* Get the path to i3 socket, or die. */
  FILE *fp;

  if (!(fp = popen("/bin/sh -c \"i3 --get-socketpath\"", "r")))
    die("popen");

  fgets(ipcaddr.sun_path, sizeof(ipcaddr.sun_path), fp);
  pclose(fp);

  if (strlen(ipcaddr.sun_path) == 0)
    die("Could not get i3 socket path");

  ipcaddr.sun_path[strlen(ipcaddr.sun_path) - 1] = 0;

  /* Subscribe to "window" events. */
  int len;
  char *payload;

  if ((sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
    die("socket");

  ipcaddr.sun_family = AF_UNIX;
  len = strlen(ipcaddr.sun_path) + sizeof(ipcaddr.sun_family);
  if (connect(sock, (struct sockaddr *)&ipcaddr, len) == -1)
    die("connect");

  // Remember about the -1 for size! Not null-terminated.
  if (send(sock, SUBSCRIBE, sizeof(SUBSCRIBE) - 1, 0) == -1)
    die("send");

  payload = get_msg(sock);

  if (payload == NULL || strlen(payload) != 16) {
    if (payload != NULL)
      free(payload);
    die("Could not subscribe for the window event");
  }

  //printf("payload: '%s'\n", payload);
  free(payload);

  /* Spawn the command. */
  pid_t pid = fork();
  if (pid < 0)
    die("fork");
  else if (pid == 0) {
    execvp(argv[ncmd], argv + ncmd);
    die("execvp");
  }

  /* Wait for nwin windows to get reparented. */
  while (nwin > 0) {
    payload = get_msg(sock);
    //printf("payload: '%s'\n", payload);
    free(payload);
    nwin--;
  }

  /* Clean up and exit. */
  shutdown(sock, 2);
  exit(0);
}
